src/writer/update-proposal.ts

```diff
@@ -0,0 +1,186 @@
+          validationName,
+          address,
+          space.id,
+          space.network,
+          'latest',
+          validationParams,
+          { url: scoreAPIUrl }
+        );
+      }
+
+      if (!isValid) return 'validation failed';
+    } catch (e) {
+      capture(e, { contexts: { input: { space: space.id, address } } });
+      log.warn(
+        `[writer] Failed to check proposal validation, ${space.id}, ${address}, ${JSON.stringify(
+          e
+        )}`
+      );
+      return 'failed to check validation';
+    }
+  }
+}
+
+export async function verify(body): Promise<any> {
+  const msg = jsonParse(body.msg);
+
+  const schemaIsValid = snapshot.utils.validateSchema(snapshot.schemas.updateProposal, msg.payload);
+  if (!schemaIsValid) {
+    log.warn('[writer] Wrong proposal format', schemaIsValid);
+    return Promise.reject('wrong proposal format');
+  }
+
+  if (!isChoicesValid(msg.payload)) {
+    return Promise.reject(`wrong choices for ${msg.payload.type} type voting`);
+  }
+
+  const proposal = await getProposal(msg.space, msg.payload.proposal);
+  if (!proposal) return Promise.reject('unknown proposal');
+
+  if (proposal.author !== body.address) return Promise.reject('Not the author');
+
+  const space = await getSpace(msg.space);
+  space.id = msg.space;
+
+  const spaceUpdateError = getSpaceUpdateError(msg.payload, space);
+  if (spaceUpdateError) return Promise.reject(spaceUpdateError);
+
+  if (isScamDetected(body.address, msg.payload))
+    return Promise.reject('scam proposal detected, contact support');
+
+  const authorizationError = await checkAuthorization(body.address, space);
+  if (authorizationError) return Promise.reject(authorizationError);
+
+  return Promise.resolve(proposal);
+}
+
+export async function action(body, ipfs): Promise<void> {
+  const msg = jsonParse(body.msg);
+  const space = msg.space;
+
+  const author = getAddress(body.address);
+  const updated = parseInt(msg.timestamp);
+  const metadata = msg.payload.metadata || {};
+  const plugins = JSON.stringify(metadata.plugins || {});
+
+  const proposal = {
+    ipfs,
+    author,
+    updated,
+    space,
+    type: msg.payload.type || 'single-choice',
+    plugins,
+    title: msg.payload.name,
+    body: msg.payload.body,
+    discussion: msg.payload.discussion || '',
+    choices: JSON.stringify(msg.payload.choices)
+  };
+
+  const query = 'UPDATE proposals SET ? WHERE id = ?';
+  const params: any[] = [proposal, msg.payload.proposal];
+
+  await db.queryAsync(query, params);
+}
+
+export function getIpfsBody(body, context) {
+  const id = context.id;
+  const originIpfs = context.ipfs;
+  const { address, sig, ...restBody } = body;
+  const { message } = restBody.data;
+
+  const updatedProposal = {
+    ...context,
+    ...message
+  };
+
+  const ipfsBody = {
+    address,
+    sig,
+    hash: id,
+    originIpfs,
+    ...restBody,
+    data: restBody.data,
+    proposal: updatedProposal
```

Comment:
Isn't `body` already the updated proposal ?

Reply:
Why do we need to store `updatedProposal` ? the create proposal writer only store the body sent, and not the final proposal

Reply:
By saving the body twice (req.body, then updatedProposal), this increase the size of the saved JSON file, which has a limit of 1MB.

This may create edge-case, where updating a proposal fail because the pinning the IPFS file fail, because it's over 1MB, whereas the content itself did not increase

Reply:
> Why do we need to store `updatedProposal` ? the create proposal writer only store the body sent, and not the final proposal

1MB is more than enough to store both entities. But anyway, what do you propose?
p.s.: you can find some considerations [here](https://discord.com/channels/955773041898573854/1149472664323379343/1149628031237885952)

Reply:
on proposal creation, it uses

```js
const { address, sig, ...restBody } = body;
const ipfsBody = {
  address,
  sig,
  hash: id,
  ...restBody
};
```

Why the same thing is not enough ? isn't `restBody` already containing the updated data ?

Reply:
it will contain only the [update body](https://github.com/snapshot-labs/snapshot.js/pull/895/files#diff-0de6060ecf3e628ed7a2ef320e36093f395de48084a4c2a4c2977588b6c0011dR142), but not the whole body of the proposal.
> isn't restBody already containing the updated data ?

no, it contains only [fields](https://github.com/snapshot-labs/snapshot.js/pull/895/files#diff-0de6060ecf3e628ed7a2ef320e36093f395de48084a4c2a4c2977588b6c0011dR142) that are allowed to update

Reply:
So I guess the reason you're storing `proposal: updatedProposal` is to have the full proposal in the IPFS ?

Reply:
Yes, + to have some kind of history with `originIpfs` key, that allows to track changes over IPFS storage

Reply:
I have removed those fields according to @bonustrack request. So I am resolving this discussion